### ISSUE TITLE
feat: Mercurial Support

### ISSUE THREAD

**Issue by nekodjin** [eza-community/eza#738]
It's all in the title. Eza already supports Git, with flags like `--git` and `--git-ignore`. This feature would extend similar support to Mercurial, with flags like `--hg` and `--hg-ignore`.

**Comment by Delapouite** [comment]
Beyond Mercurial support, lots of tools have the concept of 'ignore files': Docker, Npm, ESLint, Prettier…

So maybe we could generalize the approach with a `--ignore-file` flag with the path to `.dockerignore`, `.npmignore`… as argument to answer crucial questions like "ok what this tool currently 'see' in this directory tree?'

**Comment by MartinFillon** [comment]
Those are great ideas that could clearly be implemented.

**Comment by PThorpe92** [comment]
### Mercurial:
How popular would that feature be do you think? 

Also: It doesn't appear there is anything close to `libgit2` equivalency available (with Rust bindings anyway), although they seem to be planning to [rewrite](https://wiki.mercurial-scm.org/OxidationPlan) it in Rust so a public API would be the only shot at implementing this on par with current `git` functionality.

### other .{x}ignore files

I assume they are all parsed with the same [spec](https://git-scm.com/docs/gitignore#_pattern_format) as .gitignore files? Even if that is the case, would one flag `--ignore-files` work for all of them, including `.gitignore` ?

**Comment by MartinFillon** [comment]
Well @PThorpe92 the mercurial rust seems kinda dead, but I am trying to do a little caller to at least get the status on the side and its working pretty well its only limitation is that `hg` is in python so not the fastest thing

**Comment by nekodjin** [comment]
> Beyond Mercurial support, lots of tools have the concept of 'ignore files': Docker, Npm, ESLint, Prettier…
> 
> So maybe we could generalize the approach with a `--ignore-file` flag with the path to `.dockerignore`, `.npmignore`… as argument to answer crucial questions like "ok what this tool currently 'see' in this directory tree?'

I like the idea of an `--ignore-file` flag, but I don't think it should replace any other `--X-ignore` flags, and I think we should still consider adding more `--X-ignore` flags. `--ignore-file` is a great catch-all when a user just wants to see non-hidden files, but Eza shouldn't presume that users wouldn't want/need the granularity that they could get by combining different, _specific_ `--X-ignore` flags.

> How popular would that feature be do you think?

Git is the most popular VCS solution, and it's not even close. Having said that though, Mercurial is probably the runner-up. It's used for a number of notable projects, such as nginx. Many people think that Mercurial is a much better VCS than Git, and on an ideological level, it would be nice if tools like Eza didn't uphold the "tyranny" of Git - for the most part, the reason we all use Git is that everyone else does, and all our tools do too. The more tools support alternatives, the more likely people are to use alternatives, and the more people use alternatives, the more pressure there is for Git to catch up and get better. It's a win-win for everyone.

**Comment by PThorpe92** [comment]
>Git is the most popular VCS solution, and it's not even close. Having said that though, Mercurial is probably the runner-up

Yeah I'm not unfamiliar, I was just trying to gauge a rough estimate of how many of our users would use this feature, considering that there is no equivalent of `libgit2` that provides us with a convenient, fast API to get all of this information. 

Recently we have adopted some features/changes that have become rather difficult to maintain, and we are coming up on evaluating what will be included in a `1.0` release, and although I agree with everything you said, I'm trying to make sure we are careful about adding features at this point. 

A quick google search says Mercurial has `.032%` of the marketshare, so being generous, if `1%` of users will use this feature, we need to at least evaluate the tradeoffs. :+1:

**Comment by MartinFillon** [comment]
> A quick google search says Mercurial has .032% of the marketshare, so being generous, if 1% of users will use this feature, we need to at least evaluate the tradeoffs. 👍

Personally I dont really care being the guy that maintain something no one uses or few ones using the problem as you say is that we have 1.0 right around the corner
